store/store.go

```diff
@@ -1328,6 +1340,15 @@ func (s *Store) Download(ctx context.Context, name string, targetPath string, do
 
 	err = download(ctx, name, downloadInfo.Sha3_384, url, user, s, w, resume, pbar)
 	if err != nil {
+		// If sha3 checksum is incorrect and it was a resumed download, retry from scratch.
+		// Note that we will retry this way only once.
+		if _, ok := err.(HashError); ok && resume > 0 {
+			logger.Debugf("Error on resumed download: %v", err.Error())
+			err = os.Truncate(partialPath, 0)
+			if err == nil {
+				return s.Download(ctx, name, targetPath, downloadInfo, pbar, user)
```

Comment:
It is indeed a bit confusing, and it actually reveals that the implementation is not as simple as it could be. There's no reason for us to repeat all the boilerplate we do at the top of this function. It's only the download itself that we need to repeat. Let's use `err = w.Truncate(0)` instead of using the filepath, and let's just call the inner `download` itself directly.